### ISSUE TITLE
helper/schema: Interpolating nested list into template causes crash in ResourceData.get

### ISSUE THREAD

**Issue by 2rs2ts** [hashicorp/terraform#9488]
### Terraform Version

Terraform v0.7.7
### Affected Resource(s)
- template_file
### Terraform Configuration Files

``` hcl
provider "aws" {
  region = "us-east-1"
}

resource "aws_network_interface" "foo" {
  count = "2"
  subnet_id = "SUBNET ID GOES HERE"
}

data "template_file" "ips" {
  template = "${list(aws_network_interface.foo.*.private_ips)}"
}
```
### Debug Output

https://gist.github.com/2rs2ts/87210cf61b7610940a17de1f0c3948f0
### Panic Output

https://gist.github.com/2rs2ts/414642713e7553fe81acf2ff4b55ab86
### Expected Behavior

I'm not entirely sure what _should_ happen. I was doing this to see if I could get the plan for an `aws_route53_record` to be computed properly, see the referenced issue down below. I was hoping that terraform would mark the `records` for the plan to be a computed value. That's not exactly relevant to this particular issue though.
### Actual Behavior

Terraform crashed!
### Steps to Reproduce
1. write the above configuration to `main.tf`
2. replace the `subnet_id` with something you own
3. `terraform plan`
### References

This is me attempting a workaround for the limitations of my solution in #8696

**Comment by mitchellh** [comment]
I looked into this and can add some extra info. This is a tough one but definitely should get fixed somehow...

The main issue is that the type error isn't being caught during validation. The `template_file` data source properly defines that the `template` field is a "string." However, during validation the value is computed so it looks like a string to the validation step. By the time it reaches the plan, there is a type mismatch and there is a panic.

Plan expects that the types have been fully resolved so I believe the panic is warranted. 

What we need is for `terraform.ResourceConfig` (and perhaps `config.RawConfig`) to be able to more richly determine the final type of a computed value. It currently assumes it'll be some primitive which actually isn't too bad of an assumption until maps/lists were introduced!

**Comment by bernadinm** [comment]
I've also encountered this bug as well:

**Terraform Version**

Terraform v0.9.0

**Affected Resource(s)**

template_file

**Terraform Configuration Files**

```
variable "bootstrap_private_ip" {
 default = "1.2.3.4"
}

variable "cluster_name" {
 default = "Prod"
}

variable "masterlist" {
  type    = "list"
  default = ["admin", "ubuntu"]
}

data "template_file" "script" {
  template = "${file("run.sh")}"

   vars {
    bootstrap_private_ip = "${var.bootstrap_private_ip}"
    cluster_name         = "${var.cluster_name}"
    masterlist          = "${var.masterlist}"
 }
}
output "script" {
  value = "${data.template_file.script.rendered}"
}
```

**Panic Output**

```
terraform apply
1 error(s) occurred:

* data.template_file.script: vars (masterlist): '' expected type 'string', got unconvertible type '[]interface {}'
```

**Actual Behavior**

Terraform crashed!

**Steps to Reproduce**

1. Create a run.sh file with nothing in it
2. Create the `main.tf` file with the contents above.
3. Run `terraform apply` and see the Panic Output

I am now trying to figure out a workaround that will print out the entire list as a string instead.

**Comment by teamterraform** [comment]
Hi all! Sorry for the long silence here.

Terraform 0.12 includes a means for Terraform to track the types of unknown values, and so Terraform can now tell that the result of the expression is a list even though the exact value of that list might not be known yet. This achieves what @mitchellh was describing in his comment several years ago.

Since there's a fix for this in Terraform 0.12, we're going to close this out. If you see any other similar errors in Terraform 0.12 or later, please open a new issue for them because the configuration engine in Terraform 0.12 is now completely different and so we'll need to capture a fresh set of debug information in order to investigate.

Thanks for reporting this!
